### ISSUE TITLE
Wrong src IP when using UDP backend.

### ISSUE THREAD

**Issue by aclisp** [coreos/flannel#434]
## Tl;dr 

The `udp` backend with TUN device does not work exactly as the diagram in [README.md](https://github.com/coreos/flannel/blob/master/README.md). If a packet travels from one Pod to another, the other end can not see the correct src IP. 

Actually there is a thread in SO got the same problem:
http://stackoverflow.com/questions/34655704/flannel-tunnel-solution-in-kubernetes
## Reproduce

I have two nodes in a kubernetes cluster. The interfaces are:

|  | eth0 | docker0 | flannel0 |
| --- | --- | --- | --- |
| node1 | 10.21.8.228 | 192.168.18.1/26 | 192.168.18.0/16 |
| node2 | 10.21.8.229 | 192.168.36.1/26 | 192.168.36.0/16 |

flannel is started with default options on the two nodes. 
- docker daemon --ip-masq=true (But it does not matter, I think)
- FLANNEL_IPMASQ=false
- The backend type is udp

``` json
{
    "Network": "192.168.0.0/16",
    "SubnetLen": 26,
    "SubnetMin": "192.168.4.0",
    "SubnetMax": "192.168.254.192",
    "Backend": {
        "Type": "udp"
    }
}
```

route table on node1

```
192.168.18.0/26 dev docker0  proto kernel  scope link  src 192.168.18.1 
192.168.0.0/16 dev flannel0 
default via 10.21.8.228 dev eth0 
```

route table on node2

```
192.168.36.0/26 dev docker0  proto kernel  scope link  src 192.168.36.1 
192.168.0.0/16 dev flannel0 
default via 10.21.8.229 dev eth0
```

Kubernetes created two Pods on the two nodes. Inside the Pod, I got these information.

|  | eth0 | default route |
| --- | --- | --- |
| pod1 | 192.168.18.13/26 | via 192.168.18.1 dev eth0 |
| pod2 | 192.168.36.35/26 | via 192.168.36.1 dev eth0 |

I thought the packet from pod1 to pod2, when it arrives at pod2, should like this: 

| outer src IP | outer dst IP | inner src IP | inner dst IP |
| --- | --- | --- | --- |
| 10.21.8.228 | 10.21.8.229 | 192.168.18.13 | 192.168.36.35 |

But what captured with tcpdump at pod2's eth0 is 

| outer src IP | outer dst IP | inner src IP | inner dst IP |
| --- | --- | --- | --- |
| 10.21.8.228 | 10.21.8.229 | **192.168.18.0** | 192.168.36.35 |

The commands I used for testing:

```
pod1> ping 192.168.36.35 >/dev/null &
pod1> tcpdump -n -v icmp -i eth0

    14:02:28.555541 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto ICMP (1), length 84)
        192.168.18.13 > 192.168.36.35: ICMP echo request, id 1095, seq 21, length 64
    14:02:28.556154 IP (tos 0x0, ttl 60, id 31888, offset 0, flags [none], proto ICMP (1), length 84)
        192.168.36.35 > 192.168.18.13: ICMP echo reply, id 1095, seq 21, length 64

pod2> tcpdump -n -v icmp -i eth0

    14:02:33.555696 IP (tos 0x0, ttl 60, id 0, offset 0, flags [DF], proto ICMP (1), length 84)
        192.168.18.0 > 192.168.36.35: ICMP echo request, id 1095, seq 26, length 64
    14:02:33.555720 IP (tos 0x0, ttl 64, id 31893, offset 0, flags [none], proto ICMP (1), length 84)
        192.168.36.35 > 192.168.18.0: ICMP echo reply, id 1095, seq 26, length 64
```

**Comment by fatduo** [comment]
+1 Any update?

When a pod accept a socket connection from another, it use
> int conn = accept(server_sockfd, (struct sockaddr*)&client_addr, &length);
printf("IP: %s \n", inet_ntoa(client_addr.sin_addr));

to get the client ip address.
It gets the flannel ip instead of the pod ip.

**Comment by tomdee** [comment]
I suspect this is related to https://github.com/coreos/flannel/issues/433

**Comment by tomdee** [comment]
Ah, I see you have ip-masq off though in flannel, so maybe not.

**Comment by aclisp** [comment]
Found the root cause.

To get the correct src IP, these options are critical!
* flannel's -ip-masq=true
* docker's -ip-masq=false

In other words, flannel should insert ACCEPT rule at the head of POSTROUTING chain:
```
$ sudo iptables -t nat -S POSTROUTING
-P POSTROUTING ACCEPT
-A POSTROUTING -s 192.168.0.0/16 -d 192.168.0.0/16 -j ACCEPT
<... other MASQUERADE rules omitted>
```

In my case (the wrong src IP scenario), the head of POSTROUTING chain is something like:
```
-P POSTROUTING ACCEPT
-A POSTROUTING -s 192.168.18.64/26 ! -o docker0 -j MASQUERADE  (set by docker masq=true)
```
So every packets destined to interface `flannel0` was SNAT'ed  with `flannel0`'s IP.

**Comment by fatduo** [comment]
Hi @aclisp 
It works after I setting docker's -ip-masq=false, Thanks.
Because I'm not familiar with docker & flannel network, could you tell me the setting will cause any other problems？

**Comment by djsly** [comment]
Any expectation to have this fixed in a near 0.6.X release ?

**Comment by huggsboson** [comment]
@djsly @tomdee Isn't the fix described in @aclisp's response?  you just do ip-masq=true on flannel and ip-masq=false on docker.  Is there a change to flannel that needs to be made?

**Comment by djsly** [comment]
We went ahead and updated our init files and installer accordantly. It wasn't clear though if flannel was expecting to resolve the issue them self at one point by defaulting their config and altering the docker one. That's what I though this Issue was for. 

Sylvain

**Comment by tomdee** [comment]
Closing as I assume this is now fixed.
